### PR TITLE
 JAMES-2641 Solve concurency issues in event firing test 

### PR DESCRIPTION
**Pull request by chibenwa**
Tests proved to be unstable. After investigation, this is a Reactor related bug.

See https://github.com/linagora/james-project/pull/2144#issuecomment-458831612 and https://github.com/linagora/james-project/pull/2144#issuecomment-458810043 failues

### CODE REVIEW

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 1]
```diff
@@ -757,7 +757,7 @@ private MoveResult move(Iterator<MailboxMessage> originalRows, MailboxSession se
                     .messageId(messageIds.build())
                     .build(),
                 messageMoves.impactedMailboxIds().map(MailboxIdRegistrationKey::new).collect(Guavate.toImmutableSet())))
-            .then().block();
+            .blockLast();
```
Comment (blackheaven) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: Can you explain me why?

Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: Well... I don't know...

My method:
 - Put `@RepeatedTest(100)` every where. It breaks.
 - Play around trying to fix things -> this is my finding.

I'm also quite surprised that it is not equivalent to be honnest.

Comment (trantienduchn) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: I guess the root cause of failed tests came from `Flux.merge()` vs `Flux.concant()`, `Flux.concat()` respects the order while `Flux.merge()` doesn't

Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: Not really... all the elements in the flux were not necessarily well completed, generating that very issue.

**Comment by chibenwa** [comment]
`MemoryUploadCucumberTest » Socket Broken pipe (Write failed)` test this please

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java (hunk 1 above)
Comment (blackheaven) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: No better solution, I've tried to put `*.defer` a bit everywhere, we have a serious concurrency issue which have to be fixed.

**Comment by chibenwa** [comment]
merged

**Comment by mbaechler** [comment]
> Tests proved to be unstable. After investigation, this is a Reactor related bug.

Could you give us a link ? That's a really important topic, we have to be able to track that bug.

**Comment by chibenwa** [comment]
> Could you give us a link? That's a really important topic, we have to be able to track that bug.

I think I found it...

https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Mono.html#block--

Maybe this disclaimer cause is the root cause:

> Note that each block() will trigger a new subscription: in other words, the result might miss signal from hot publishers.

**Comment by mbaechler** [comment]
I doesn't sound related, we don't have a hot publisher, don't we ?

**Pull request closed without merging**